core/trino-main/src/main/java/io/trino/sql/planner/AdaptivePlanner.java

```diff
@@ -0,0 +1,438 @@
+ * RemoteSourceNode for finished stages will remain as it is in the plan.
+ *
+ * 2. Once we have a merged plan which contains all the unfinished parts, we will reoptimize it using
+ * a set of PlanOptimizers.
+ *
+ * 3. During re-optimization, it is possible that some new exchanges need to be added due to the change
+ * in partitioning strategy. For instance, if a rule changes the distribution type of the join
+ * from BROADCAST to PARTITIONED. It is also possible that some remote exchanges are removed.
+ * For example, while changing the order of the join.
+ *
+ * 4. Ultimately, the planner will fragment the optimized PlanNode again and generate the SubPlans with
+ * new PlanFragmentIds. The re-fragmentation will only happen if the old plan and the new
+ * plan have some differences. To check these differences, we rely on
+ * PlanOptimizer#optimizeAndReturnChanges api which also returns changes plan ids.
+ *
+ * Note: We do not change the fragment ids which have no changes and are not downstream of the
+ * changed plan nodes. This optimization is done to avoid unnecessary stage restart due to speculative execution.
+ */
+public class AdaptivePlanner
+{
+    private final Session session;
+    private final PlannerContext plannerContext;
+    private final List<AdaptivePlanOptimizer> planOptimizers;
+    private final PlanFragmenter planFragmenter;
+    private final PlanSanityChecker planSanityChecker;
+    private final IrTypeAnalyzer typeAnalyzer;
+    private final WarningCollector warningCollector;
+    private final PlanOptimizersStatsCollector planOptimizersStatsCollector;
+
+    public AdaptivePlanner(
+            Session session,
+            PlannerContext plannerContext,
+            List<AdaptivePlanOptimizer> planOptimizers,
+            PlanFragmenter planFragmenter,
+            PlanSanityChecker planSanityChecker,
+            IrTypeAnalyzer typeAnalyzer,
+            WarningCollector warningCollector,
+            PlanOptimizersStatsCollector planOptimizersStatsCollector)
+    {
+        this.session = requireNonNull(session, "session is null");
+        this.plannerContext = requireNonNull(plannerContext, "plannerContext is null");
+        this.planOptimizers = requireNonNull(planOptimizers, "planOptimizers is null");
+        this.planFragmenter = requireNonNull(planFragmenter, "planFragmenter is null");
+        this.planSanityChecker = requireNonNull(planSanityChecker, "planSanityChecker is null");
+        this.typeAnalyzer = requireNonNull(typeAnalyzer, "typeAnalyzer is null");
+        this.warningCollector = requireNonNull(warningCollector, "warningCollector is null");
+        this.planOptimizersStatsCollector = requireNonNull(planOptimizersStatsCollector, "planOptimizersStatsCollector is null");
+    }
+
+    public SubPlan optimize(SubPlan root, StageRuntimeInfoProvider stageRuntimeInfoProvider)
+    {
+        // No need to run optimizer since the root is already finished or its stats are estimated by progress.
```

Comment:
The logic here is not obvious to me. Why are we treating "estimated by progress" as something special. 
What I think should be happening here is that we estimate how much work is there still remaining to be done with current plan, and compare that with how much work the new plan would do. And only commit to optimization if latter is smaller.